controller/src/main/java/org/jboss/as/controller/AttributeDefinition.java

```diff
@@ -108,12 +111,57 @@ protected AttributeDefinition(String name, String xmlName, final ModelNode defau
 
         this(name, xmlName, defaultValue, type, allowNull, allowExpression, measurementUnit, valueCorrector,
                 wrapValidator(validator, allowNull, validateNull, allowExpression, type), validateNull, alternatives, requires,
                 attributeMarshaller, resourceOnly, deprecationData, wrapConstraints(accessConstraints),
-                nilSignificant, parser, null, null, null, wrapFlags(flags));
+                nilSignificant, parser, null, null, null, undefinedMetricValue, wrapFlags(flags));
+    }
+
+    private AttributeDefinition(String name, String xmlName, final ModelNode defaultValue, final ModelType type,
+                                final boolean allowNull, final boolean allowExpression, final MeasurementUnit measurementUnit,
+                                final ParameterCorrector valueCorrector, final ParameterValidator validator, final boolean validateNull,
+                                final String[] alternatives, final String[] requires, AttributeMarshaller attributeMarshaller,
+                                boolean resourceOnly, DeprecationData deprecationData, final List<AccessConstraintDefinition> accessConstraints,
+                                Boolean nilSignificant, AttributeParser parser, final String attributeGroup, CapabilityReferenceRecorder referenceRecorder,
+                                ModelNode[] allowedValues, final ModelNode undefinedMetricValue, final EnumSet<AttributeAccess.Flag> flags) {
+
+        this.name = name;
+        this.xmlName = xmlName == null ? name : xmlName;
+        this.type = type;
+        this.allowNull = allowNull;
+        this.allowExpression = allowExpression;
+        this.parser = parser != null ? parser : AttributeParser.SIMPLE;
+        this.defaultValue = new ModelNode();
+        if (defaultValue != null) {
+            this.defaultValue.set(defaultValue);
+        }
+        this.defaultValue.protect();
+        this.measurementUnit = measurementUnit;
+        this.alternatives = alternatives;
+        this.requires = requires;
+        this.valueCorrector = valueCorrector;
+        this.validator = validator;
+        this.validateNull = validateNull;
+        this.flags = flags;
+        if (attributeMarshaller != null) {
+            this.attributeMarshaller = attributeMarshaller;
+        } else {
+            this.attributeMarshaller = new DefaultAttributeMarshaller();
+        }
+        this.resourceOnly = resourceOnly;
+        this.accessConstraints = accessConstraints;
+        this.deprecationData = deprecationData;
+        this.nilSignificant = nilSignificant;
+        this.attributeGroup = attributeGroup;
+        this.allowedValues = allowedValues;
+        this.undefinedMetricValue = new ModelNode();
+        if (undefinedMetricValue != null) {
+            this.undefinedMetricValue.set(undefinedMetricValue);
+        }
```

Comment:
getUndefinedMetricValue() is going to return null if the constructor param is null or undefined, so in that case we can skip constructing a ModelNode and just store null. Same applies for the defaultValue treatment above (which I wrote.)

I've been working a managment perf support case, so I have perf on my mind today. ;)